File: Simulation C++/Grid.cpp
#include <iostream>
#include <fstream>
#include <cmath>
#include <string>
#include<assert.h>
#include "Grid.h"
using namespace std;

Grid::Grid(int nn){
    n = nn;
	std::cout << n << "\n";
    x = new double[n];
    conc = new double[4*n];
	concA = new double[n];
	concB = new double[n];
	concY = new double[n];
	concZ = new double[n];
	g = 0.0;
	gB = 0.0;
}

Grid::~Grid(){
    delete[] x;
    delete[] conc;
}

void Grid::grid(double dx, double gamma){
    x[0] = 0.0;
    for(int i = 1; i < n; i++){
        x[i] = x[i-1] + dx;
    }
}

/*void Grid::init_c(double sc){
    for(int i = 0; i < n; i++){
        conc[i] = sc;
    }
}*/

void Grid::init_c(double A, double B, double Y,double Z) {
	for (int i = 0; i < 4 * n; i = i + 4) {
		conc[i] = A;
		conc[i + 1] = B;
		conc[i + 2] = Y;
		conc[i + 3] = Z;
	}
	for (int i = 0; i < n; i++) {
		concA[i] = A;
		concB[i] = B;
		concY[i] = Y;
		concZ[i] = Z;
	}
}

void Grid::grad(){
    g = -(conc[5] - conc[1]) / (x[1] - x[0]);
	std::cout << conc[5] << "\t" << conc[1] << "\t" << x[1] << "\t" << x[0]<< "\n";
}
void Grid::gradB() {
	gB = (conc[5] - conc[1]) / (x[1] - x[0]);
}
void Grid::updateAll() {
	for (int j = 0, i = 0; j <n*4; j = j + 4, i++) {
		concA[i] = conc[j];
		concB[i] = conc[j + 1];
		concY[i] = conc[j + 2];
		concZ[i] = conc[j + 3];

	}
	/*for (int i = 3; i < n * 3 - 3; i++) {
		assert(conc[i] <= 1.000000001 && conc[i] >= -0.001 && "Concentration of species out of bound");
		if (i >= 6 && i < n - 3) {
			assert(fabs(conc[i] - conc[i - 3]) < 0.1 && "Large variation with previous concentrtaion");
			assert(fabs(conc[i] - conc[i + 3]) < 0.1 && "Large variation from next concentration");
		}
	}*/
}

void Grid::saveA(string filename)
{
    ofstream file(filename.c_str(), ios::out);
    for(int i = 0; i < n; i++)
    {
        file << x[i] << "," << concA[i] << endl;
    }
    file.close();
}
void Grid::saveB(string filename)
{
	ofstream file(filename.c_str(), ios::out);
	for (int i = 0; i < n; i++)
	{
		file << x[i] << "," << concB[i] << endl;
	}
	file.close();
}
void Grid::saveY(string filename)
{
	ofstream file(filename.c_str(), ios::out);
	for (int i = 0; i < n; i++)
	{
		file << x[i] << "," << concY[i] << endl;
	}
	file.close();
}

void Grid::saveZ(string filename)
{
	ofstream file(filename.c_str(), ios::out);
	for (int i = 0; i < n; i++)
	{
		file << x[i] << "," << concZ[i] << endl;
	}
	file.close();
}

File: Simulation Python/grid.py
import numpy as np
import csv

class Grid(object):
    def __init__(self,n):
        self.n = n
        self.x = np.zeros(self.n,dtype=np.float64)
        self.conc = np.zeros(self.n*4,dtype=np.float64)
        self.concA = np.zeros(self.n,dtype=np.float64)
        self.concB = np.zeros(self.n,dtype=np.float64)
        self.concY = np.zeros(self.n,dtype=np.float64)
        self.concZ = np.zeros(self.n,dtype=np.float64)

        self.g = 0.0


    def grid(self,dX,gamma):
        self.x[0] = 1.0
        for i in range(1,self.n):
            self.x[i] = self.x[i-1] + dX
            dX = dX* (1.0 +gamma )

    
    # initialize the concentration matrix
    def init_c(self,A:float,B:float,Y:float,Z:float,Theta:float):
        self.conc[::4] = A
        self.conc[1::4] = B
        self.conc[2::4] = Y
        self.conc[3::4] = Z

        self.concA[:] = A
        self.concB[:] = B
        self.concY[:] = Y 
        self.concZ[:] = Z

    # A better initialized concentration vector
    
    
    
    """
    def init_c(self,A:float,B:float,Y:float,Z:float,Theta:float):
        NernstB = B*1.0/(1.0 + np.exp(-Theta))
        NernstY = B*1.0/(1.0 + np.exp(Theta))

        self.concB = np.linspace(NernstB,B,num=self.n,endpoint=True)
        self.concY = np.linspace(NernstY,Y,num=self.n,endpoint=True)
        self.concZ[:] = Z
        self.concA[:] = A

        self.conc[::4] = self.concA
        self.conc[1::4] = self.concB
        self.conc[2::4] = self.concY
        self.conc[3::4] = self.concZ

        #print(self.conc)
    """
    """
    
    def init_c(self,A:float,B:float,Y:float,Z:float,Theta:float):
        NernstB = B*1.0/(1.0 + np.exp(-Theta))
        NernstY = B*1.0/(1.0 + np.exp(Theta))

        self.conc[::4] = A
        self.conc[1::4] = B
        self.conc[2::4] = Y
        self.conc[3::4] = Z

        self.conc[1] = NernstB
        self.conc[2] = NernstY

        self.concA = self.conc[::4]
        self.concB = self.conc[1::4]
        self.concY = self.conc[2::4]
        self.concZ = self.conc[3::4]  

    """
    def grad(self):
        self.g = -(self.conc[5]-self.conc[1]) / (self.x[1]-self.x[0])

        return self.g

    def updateAll(self):
        self.concA = self.conc[::4]
        self.concB = self.conc[1::4]
        self.concY = self.conc[2::4]
        self.concZ = self.conc[3::4]

    def saveA(self,filename):
        f=open(filename,mode='w',newline='')
        writer = csv.writer(f)
        for i in range(self.n):
            writer.writerow([self.x[i],self.concA[i]])
        f.close()

    def saveB(self,filename):
        f=open(filename,mode='w',newline='')
        writer = csv.writer(f)
        for i in range(self.n):
            writer.writerow([self.x[i],self.concB[i]])
        f.close()
    
    def saveY(self,filename):
        f=open(filename,mode='w',newline='')
        writer = csv.writer(f)
        for i in range(self.n):
            writer.writerow([self.x[i],self.concY[i]])
        f.close()
    
    def saveZ(self,filename):
        f=open(filename,mode='w',newline='')
        writer = csv.writer(f)
        for i in range(self.n):
            writer.writerow([self.x[i],self.concZ[i]])
        f.close()


    def massConservation(self,A,B,Y,Z):
        massInitial = 0.0
        massA = 0.0
        massB = 0.0
        massY = 0.0
        massZ = 0.0 
        tempA = 0.0
        tempB = 0.0
        tempY = 0.0
        tempZ = 0.0
        for i in range(self.n-1):
            h = self.x[i+1] - self.x[i]
            massInitial += h*(2*A+B+Y+Z)
            tempA = (self.concA[i] + self.concA[i + 1]) / 2.0
            massA += 2.0*tempA * h
            tempB = (self.concB[i] + self.concB[i + 1]) / 2.0
            massB += tempB * h
            tempY = (self.concY[i] + self.concY[i + 1]) / 2.0
            massY += tempY * h
            tempZ = (self.concZ[i] + self.concZ[i + 1]) / 2.0
            massZ += tempZ * h

        massFinal = massA + massB + massY + massZ

        return (massFinal-massInitial)/massInitial